### PR TITLE
cmake: Add `libatomic` check

### PR DESCRIPTION
**Pull request by hebasto**
1. Added an analogue of https://github.com/bitcoin/bitcoin/blob/master/build-aux/m4/l_atomic.m4.
2. Included changes from https://github.com/bitcoin/bitcoin/pull/28777.
3. Added (temporarily?) a test case to the GHA CI to avoid regressions in the future.

Might be tested, for example, as follows:
```
make -j $(nproc) -C depends HOST=i686-pc-linux-gnu CC="clang -m32" CXX="clang++ -m32" NO_QT=1 NO_WALLET=1 NO_ZMQ=1 NO_NATPMP=1 NO_UPNP=1 NO_USDT=1
cmake -B build --toolchain depends/i686-pc-linux-gnu/share/toolchain.cmake
cmake --build build -j $(nproc) -t bitcoind
```

Here is an excerpt from the CI [log](https://github.com/hebasto/bitcoin/actions/runs/6842095586/job/18603196018):
```
-- Performing Test STD_ATOMIC_LINKS_WITHOUT_LIBATOMIC
-- Performing Test STD_ATOMIC_LINKS_WITHOUT_LIBATOMIC - Failed
-- Performing Test STD_ATOMIC_NEEDS_LINK_TO_LIBATOMIC
-- Performing Test STD_ATOMIC_NEEDS_LINK_TO_LIBATOMIC - Success
```

### CODE REVIEW

**Comment by hebasto** [comment]
Friendly ping @fanquake @theuni @TheCharlatan @vasild @pablomartin4btc :)

File: cmake/module/WarnAboutGlobalProperties.cmake [hunk 1]
```diff
@@ -22,3 +22,12 @@ get_directory_property(global_link_options LINK_OPTIONS)
...
   message(AUTHOR_WARNING "The directory's LINK_OPTIONS property is not empty: ${global_link_options}")
 endif()
+
+# Avoid the directory-wide link_libraries() command.
+# Instead, prefer the target-specific target_link_libraries() one.
+file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/dummy_cxx_source.cpp "#error")
+add_library(check_loose_linked_libraries OBJECT EXCLUDE_FROM_ALL ${CMAKE_CURRENT_BINARY_DIR}/dummy_cxx_source.cpp)
+get_target_property(global_linked_libraries check_loose_linked_libraries LINK_LIBRARIES)
+if(global_linked_libraries)
+  message(AUTHOR_WARNING "There are libraries linked with `link_lbraries` commands: ${global_linked_libraries}")
```
Comment (vasild) [inline cmake/module/WarnAboutGlobalProperties.cmake:32]: typo: `s/link_lbraries/link_libraries/`

**Review by vasild — APPROVED** [review]
LGTM c8c4c6280b

File: cmake/module/WarnAboutGlobalProperties.cmake (hunk 1 above)
Comment (hebasto) [reply cmake/module/WarnAboutGlobalProperties.cmake]: Thanks! Fixed.

File: cmake/introspection.cmake [hunk 2]
```diff
@@ -75,29 +75,9 @@ check_include_file_cxx(ifaddrs.h HAVE_IFADDRS_H)
...
-    if(NOT IFADDR_LINKS_WITHOUT_LIBSOCKET)
-      check_cxx_source_links_with_libs(socket "${check_socket_source}" IFADDR_NEEDS_LINK_TO_LIBSOCKET)
-      if(IFADDR_NEEDS_LINK_TO_LIBSOCKET)
-        link_libraries(socket)
-      else()
-        message(FATAL_ERROR "Cannot figure out how to use getifaddrs/freeifaddrs.")
-      endif()
-    endif()
+    include(TestAppendRequiredLibraries)
+    test_append_socket_library(core)
```
Comment (theuni) [inline cmake/introspection.cmake:80]: Being pedantic, this shouldn't be `core`, and it shouldn't be globally scoped upstream. I think it's fine here as-is, but we should plan to break this out of `core` at some point in the future.

In more detail: I think it's actually a minority of our binaries that requires `getifaddrs`. bitcoind obviously requires it, but libbitcoinkernel and bitcoin-chainstate shouldn't.

File: cmake/module/WarnAboutGlobalProperties.cmake [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2023-present The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or https://opensource.org/license/mit/.
+
```
Comment (theuni) [inline cmake/module/WarnAboutGlobalProperties.cmake:4]: Need include_guard ?

**Review by theuni — COMMENTED** [review]
Generally LGTM.

File: cmake/module/WarnAboutGlobalProperties.cmake (hunk 3 above)
Comment (hebasto) [reply cmake/module/WarnAboutGlobalProperties.cmake:4]: Thanks! Done.

File: cmake/introspection.cmake (hunk 2 above)
Comment (hebasto) [reply cmake/introspection.cmake:80]: > I think it's fine here as-is, but we should plan to break this out of core at some point in the future.

Agree.

**Review by theuni — APPROVED** [review]
ACK

**Pull request merged**